File: day17/day17.cpp
#include <algorithm>
#include <iostream>
#include <string>
#include <sstream>
#include <list>
#include <iterator>
#include <cstring>
#include <fstream>
#include <functional>
#include <cassert>

using namespace std;

struct Result {
    int value_after_last_insert;
    int value_after_zero;
};

Result play(int stepsize, int rounds) {
    list<int> buff = { 0 };
    auto pos = buff.begin();
    for (int i = 1; i <= rounds; ++i) {
        if (i % 100000 == 0) {
            cout << "round " << i << endl;
        }
        for(int i = 0; i < stepsize; ++i) {
            ++pos;
            if(pos == buff.end()) {
                pos = buff.begin();
            }
        }
        pos = buff.insert(++pos, i);
        /* cout << "insert " << i << endl; */
        /* for (auto v : buff) { */
        /*     cout << v << ", "; */
        /* } */
        /* cout << endl; */
    }
    ++pos;
    if(pos == buff.end()) {
        pos = buff.begin();
    }
    Result result;
    result.value_after_last_insert = *pos;
    result.value_after_zero = *(++find(buff.begin(), buff.end(), 0));
    return result;
}

int main() {
    assert(play(3, 2017).value_after_last_insert == 638);
    Result result1 = play(316, 2017);
    cout << "part 1 = " << result1.value_after_last_insert << endl;
    Result result2 = play(316, 50000000);
    cout << "part 2 = " << result2.value_after_zero << endl;
}

File: day17/day17.py
#! /usr/bin/env python3

def main():
    assert play(3, 2017)[0] == 638
    result1, _ = play(316, 2017)
    print("part 1 =", result1)
    _, result2 = play(316, 50000000)
    print("part 2 =", result)

def play(stepsize, rounds):
    buff = [0]
    pos = 0
    for i in range(rounds):
        if i % 100000 == 0:
            print("round", i)
        pos = ((pos + stepsize) % len(buff)) + 1
        buff.insert(pos, i+1)
    return buff[pos+1], buff[buff.index(0)+1]

if __name__ == "__main__":
    main()
